community/kernel/src/main/java/org/neo4j/kernel/impl/api/schema/BridgingIndexProgressor.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api.schema;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import org.neo4j.internal.kernel.api.IndexQuery;
+import org.neo4j.kernel.api.schema.index.IndexDescriptor;
+import org.neo4j.storageengine.api.schema.IndexProgressor;
+import org.neo4j.values.storable.Value;
+
+/**
+ * Combine multiple progressor to act like one single logical progressor seen from client's perspective.
+ */
+public class BridgingIndexProgressor implements IndexProgressor.NodeValueClient, IndexProgressor
+{
+    private final NodeValueClient client;
+    private final int[] keys;
+    // This is a thread-safe queue because it can be used in parallel scenarios.
+    // The overhead of a concurrent queue in this case is negligable since typically there will be two or a very few number
+    // of progressors and each progressor has many results each
+    private final Queue<IndexProgressor> progressors = new ConcurrentLinkedQueue<>();
```

Comment:
What is the parallel scenario?

Reply:
Querying a lucene index that has multiple partitions is done in parallel.